src/hotspot/share/gc/g1/g1ConcurrentMark.cpp

```diff
@@ -1102,6 +1102,16 @@ void G1ConcurrentMark::remark() {
     return;
   }
 
+  // collect the liveness info from all referenced regions
+  // the info should be consistent form the previous marking phase
+  size_t live_size = 0;
+  uint max_reserved_regions = _g1h->max_reserved_regions();
+  for (uint i = 0; i < max_reserved_regions; i++) {
+    live_size += _region_mark_stats[i]._live_words;
+  }
+
+  _g1h->set_live(live_size * HeapWordSize);
+
```

Comment:
This code is located in the wrong place. It will return only the live words for the areas that have been marked, not eden or objects allocated in old gen after the marking started.

Further it iterates over all regions, which can be large compared to actually active regions.

A better place is in `G1UpdateRemSetTrackingBeforeRebuild::do_heap_region()` after the last method call - at that point, `HeapRegion::live_bytes()` contains the per-region number of live data for all regions.

`G1UpdateRemSetTrackingBeforeRebuild` is instantiated and then called by multiple threads. It's probably best that that `HeapClosure` locally sums up the live byte estimates and then in the caller `G1UpdateRemSetTrackingBeforeRebuildTask::work()` sums up the per thread results like is done for `G1UpdateRemSetTrackingBeforeRebuildTask::_total_selected_for_rebuild`, which is then set in the caller of the `G1UpdateRemSetTrackingBeforeRebuildTask`.

Reply:
Would something along the line of https://github.com/openjdk/jdk/pull/2579/commits/08c715abccddbd04ced58706b7a705670843b43a be ok?